### PR TITLE
Improve locale state in calypso

### PR DESCRIPTION
**Pull request by Tug**
This PR seeks to improve how locale is set or updated by moving the value from the context to the redux state.

#### Testing instructions
  
1. Run `git checkout update/lang` and start your server, or open a [live branch](https://calypso.live/?branch=update/lang)
2. Open the [`Login` page](http://calypso.localhost:3000/log-in/fr)
3. Assert that you see the correct language & no CSS is loaded twice
4. Assert that correct (RTL/LTR) CSS tags are received from the server ( view-source )

| User logged in | User language | URL Language Param | Result language |
|----------------|---------------|--------------------|-----------------|
| ✓              | HE            | FR                 | FR              |
| ✓              | HE            | HE                 | HE              |
| ✓              | FR            | FR                 | FR              |
| ✓              | HE            | N/A                | HE              |
| ✓              | FR            | N/A                | FR              |
| ✗              | N/A           | HE                 | HE              |
| ✗              | N/A           | FR                 | FR              |
| ✗              | N/A           | N/A                | EN              |
 
5. Assert that the Support User feature still works as intended and that you are switched to the user's locale.

* Note: Each test should be performed with and without `"wpcom-user-bootstrap": true`. See #18264 
#### Additional notes

* Commit [d54d453](https://github.com/Automattic/wp-calypso/pull/18297/commits/d54d4533df989969df678132195dbae6c1e5e8fd) is only for testing and should be removed before merging.
* SSR is only used for the default locale ( EN ).

#### Reviews
  
- [x] Code
- [x] Product
- [ ] Tests

### CODE REVIEW

File: client/state/selectors/is-rtl.js [hunk 1]
```diff
@@ -10,9 +15,7 @@ import { getCurrentUser } from 'state/current-user/selectors';
  * @return {?Boolean}            Current user is rtl
  */
 export default function isRtl( state ) {
-	const currentUser = getCurrentUser( state );
-	if ( currentUser && currentUser.hasOwnProperty( 'isRTL' ) ) {
-		return currentUser.isRTL;
-	}
-	return null;
+	const localeSlug = get( state, 'ui.language.localeSlug', 'en' );
```
Comment (yurynix) [inline client/state/selectors/is-rtl.js]: Should use selector `getCurrentLocaleSlug`

File: client/state/selectors/get-current-locale-slug.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+ */
+import createSelector from 'lib/create-selector';
+
+/**
+ * Gets the current ui locale slug
+ * @param {Object} state - global redux state
+ * @return {String} current state value
+ */
+export const getCurrentLocaleSlug = createSelector(
+	state => state.ui.language.localeSlug
```
Comment (yurynix) [inline client/state/selectors/get-current-locale-slug.js]: Should use `get()`

File: client/state/ui/language/actions.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+} from 'state/action-types';
+
+/**
+ * Set the ui locale
+ *
+ * @param   {Object} localeSlug the locale slug to change the locale to
+ * @returns {Function} Action thunk
+ */
+export const setLocale = localeSlug => {
+	switchLocale( localeSlug );
```
Comment (yurynix) [inline client/state/ui/language/actions.js]: SSR ignores everything that's not the default locale:
https://github.com/Automattic/wp-calypso/blob/master/server/render/index.js#L95

Also `i18n` lib is a singleton, so it's state will be shared between all the requests,
calling `i18n.setLocale` on server side won't get us the result we want.

So I'm not sure if we even need to call it on the server.

File: client/controller/shared.js [hunk 4]
```diff
@@ -109,7 +109,7 @@ export function setUpLocale( context, next ) {
 	} else if ( currentUser ) {
 		context.lang = currentUser.localeSlug;
 	} else {
-		context.lang = context.lang || config( 'i18n_default_locale_slug' );
+		context.lang = config( 'i18n_default_locale_slug' );
```
Comment (yurynix) [inline client/controller/shared.js]: Why do we even set `context.lang` if we overwrite it later at `server/render/index.js`:
```js
	if ( context.store ) {
        ...
		context.isRTL = isRTL( context.store.getState() );
		context.lang = getCurrentLocaleSlug( context.store.getState() );
	}
```

Comment (yurynix) [reply client/controller/shared.js]: I also don't think we should check here for current user, this is already checked on boot,
so IMO the only code needs to be here is so:

```js
export function setUpLocale( context, next ) {
   if ( context.params.lang ) {
      context.store.dispatch( setLocale( context.params.lang ) );
   }
   
 	loadSectionCSS( context, next ); // not sure about that
}
```

File: server/render/index.js [hunk 5]
```diff
@@ -104,18 +104,23 @@ export function serverRender( req, res ) {
 		metas = getDocumentHeadMeta( context.store.getState() );
 		links = getDocumentHeadLink( context.store.getState() );
 
-		let reduxSubtrees = [ 'documentHead' ];
+		const cacheableReduxSubtrees = [ 'documentHead' ];
+		let reduxSubtrees;
 
 		// Send redux state only in logged-out scenario
```
Comment (yurynix) [inline server/render/index.js]: The comment is no longer true with that code,
whether we're logged in or logged out is not checked.

I'm not sure we need to remove `! context.user`, what if `ui` subtree contains some user specific state? Like the state of a credit card input field?

**Comment by yurynix** [comment]
`http://calypso.localhost:3000/log-in/fr`displaying Hebrew if my user is set to Hebrew.

**Comment by yurynix** [comment]
When using user-bootstrapping, LTR css is served. Investigating...

File: client/notifications/index.jsx [hunk 6]
```diff
@@ -153,7 +148,7 @@ export class Notifications extends Component {
 	};
 
 	render() {
-		const localeSlug = get( user.get(), 'localeSlug', config( 'i18n_default_locale_slug' ) );
+		const localeSlug = get( this.props.currentLocaleSlug, 'localeSlug', config( 'i18n_default_locale_slug' ) );
```
Comment (Tug) [inline client/notifications/index.jsx]: This looks wrong

File: client/boot/common.js [hunk 7]
```diff
@@ -23,12 +23,16 @@ import {
...
+import getCurrentLocaleSlug from 'state/selectors/get-current-locale-slug';
 
 const debug = debugFactory( 'calypso' );
 
 const switchUserLocale = ( currentUser, reduxStore ) => {
+	const currentLocaleSlug = getCurrentLocaleSlug( reduxStore.getState() );
 	const localeSlug = currentUser.get().localeSlug;
-	if ( localeSlug ) {
+
+	if ( isDefaultLocale( currentLocaleSlug ) && localeSlug ) {
```
Comment (Tug) [inline client/boot/common.js]: Why limit to only switching to english? that looks quite limited

File: server/pages/index.js [hunk 8]
```diff
@@ -180,8 +180,7 @@ function getDefaultContext( request ) {
 		// use ipv4 address when is ipv4 mapped address
 		clientIp: request.ip ? request.ip.replace( '::ffff:', '' ) : request.ip,
 		isDebug: context.env === 'development' || context.isDebug,
-		tinymceWpSkin: context.urls[ 'tinymce/skins/wordpress/wp-content.css' ],
-		tinymceEditorCss: context.urls[ 'editor.css' ]
+		urls: context.urls,
```
Comment (Tug) [inline server/pages/index.js]: `context.urls` is filled with the urls of every generated chunks. It's probably not required for now

**Comment by Tug** [comment]
Did some minor fixes, rebased and tested, I think it's ready to go.

Tests are failing with `TypeError: (0 , _config2.default)(...).reduce is not a function` here https://github.com/Automattic/wp-calypso/pull/18297/files#diff-138fcb256fa5ea3ba188a240f6f007eeR14 but I'm not sure what's wrong. Looks like jest is doing some mocking or something. @gziolo perhaps?

**Comment by gziolo** [comment]
It looks like optimization introduced in this PR doesn't play well with the all existing mocks that were introduced to drive tests using config. I added a function which postpones config access until it is really necessary. It is still possible to add logic which uses a reducer, but I wanted in the first place to find out what the issue is.

```diff
diff --git a/client/lib/i18n-utils/utils.js b/client/lib/i18n-utils/utils.js
index 65eb727499..feb3624345 100644
--- a/client/lib/i18n-utils/utils.js
+++ b/client/lib/i18n-utils/utils.js
@@ -1,3 +1,4 @@
+/** @format */
 /**
  * External dependencies
  */
@@ -11,10 +12,8 @@ import config from 'config';
 const localeRegex = /^[A-Z]{2,3}$/i;
 const localeWithRegionRegex = /^[A-Z]{2,3}-[A-Z]{2,3}$/i;
 
-const languagesMap = config( 'languages' ).reduce( ( result, language ) => {
-       result[ language.langSlug ] = language;
-       return result;
-}, {} );
+const findLanguage = langSlug =>
+       config( 'languages' ).find( language => language.langSlug === langSlug );
 
 function getPathParts( path ) {
        // Remove trailing slash then split. If there is a trailing slash,
@@ -37,7 +36,7 @@ const i18nUtils = {
                let language;
 
                if ( localeRegex.test( langSlug ) || localeWithRegionRegex.test( langSlug ) ) {
-                       language = languagesMap[ langSlug ] || languagesMap[ langSlug.split( '-' )[ 0 ] ];
+                       language = findLanguage( langSlug ) || findLanguage( langSlug.split( '-' )[ 0 ] );
                }
 
                return language;
```

**Pull request merged**